torch/csrc/api/src/nn/modules/dropout.cpp

```diff
@@ -8,42 +9,71 @@
 #include <ostream>
 #include <vector>
 
+namespace F = torch::nn::functional;
+
 namespace torch {
 namespace nn {
-namespace detail {
-template <typename Derived>
-DropoutImplBase<Derived>::DropoutImplBase(const DropoutOptions& options_)
-    : options(options_) {
-  TORCH_CHECK(options.rate() >= 0, "Dropout rate must not be less than zero");
-  TORCH_CHECK(options.rate() <= 1, "Dropout rate must not be greater than one");
+
+Tensor DropoutImpl::forward(Tensor input) {
+  return F::detail::dropout(input, options.p(),
+      is_training(), options.inplace());
 }
 
-template <typename Derived>
-void DropoutImplBase<Derived>::reset() {}
+void DropoutImpl::pretty_print(std::ostream& stream) const {
+  stream << std::boolalpha
+         << "torch::nn::Dropout(p=" << options.p()
+         << ", inplace=" << options.inplace() << ")";
+}
 
-template class DropoutImplBase<DropoutImpl>;
-template class DropoutImplBase<FeatureDropoutImpl>;
-} // namespace detail
+// ============================================================================
 
-DropoutImpl::DropoutImpl(const DropoutOptions& options_) : DropoutImplBase(options_) {}
+Tensor Dropout2dImpl::forward(Tensor input) {
+  return F::detail::dropout2d(input, options.p(),
+      is_training(), options.inplace());
+}
 
-Tensor DropoutImpl::forward(const Tensor& input) {
-  return torch::dropout(input, options.rate(), this->is_training());
+void Dropout2dImpl::pretty_print(std::ostream& stream) const {
+  stream << std::boolalpha
+         << "torch::nn::Dropout2d(p=" << options.p()
+         << ", inplace=" << options.inplace() << ")";
 }
 
-void DropoutImpl::pretty_print(std::ostream& stream) const {
-  stream << "torch::nn::Dropout(rate=" << options.rate() << ")";
+// ============================================================================
+
+Tensor Dropout3dImpl::forward(Tensor input) {
+  return F::detail::dropout3d(input, options.p(),
+      is_training(), options.inplace());
 }
 
-FeatureDropoutImpl::FeatureDropoutImpl(const DropoutOptions& options_)
-    : DropoutImplBase(options_) {}
+void Dropout3dImpl::pretty_print(std::ostream& stream) const {
+  stream << std::boolalpha
+         << "torch::nn::Dropout3d(p=" << options.p()
+         << ", inplace=" << options.inplace() << ")";
+}
+
+// ============================================================================
+
+FeatureDropoutImpl::FeatureDropoutImpl(double p)
+  : detail::_DropoutNd<FeatureDropoutImpl>(p) {
+  TORCH_WARN("torch::nn::FeatureDropout module is deprecated."
+             "Use Dropout{2,3}d instead.");
+}
+
+FeatureDropoutImpl::FeatureDropoutImpl(const FeatureDropoutOptions& options_)
+  : detail::_DropoutNd<FeatureDropoutImpl>(options_) {
+  TORCH_WARN("torch::nn::FeatureDropout module is deprecated."
+             "Use Dropout{2,3}d instead.");
+}
```

Comment:
It would be awesome to test for the deprecation message, similar to https://github.com/pytorch/pytorch/blob/eef349a679f5331e05de3c9462e285d1ab99c8a8/test/cpp/api/modules.cpp#L1176-L1188